# STAT 207 Lab Assignment 6 - Group Part - 7 Points


## <u>Case Study</u>: Reviewing Most Missed Concepts from the Bonus Quiz

In this group assignment we'll review some of the concepts that were most missed in Tuesday's **Midterm 1 Bonus Point Review Quiz.**

<p>&nbsp;</p>
<table style="border: none;border-collapse: collapse;width:130pt;">
    <tbody>
        <tr>
            <td style="color:black;font-size:15px;font-weight:700;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;width:59pt;">Problem</td>
            <td style="color:black;font-size:15px;font-weight:700;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:general;vertical-align:bottom;border:.5pt solid windowtext;border-left:none;width:71pt;">Points</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">1.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.25</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">1.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">1.3</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">1.4</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">1.5</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">3.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.25</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">3.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">3.3.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">3.3.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">3.3.3</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1</td>
        </tr>
    </tbody>
</table>

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt


**Contribution Report**

Ava Zhang and Nora Harvey - 2
Syed Ammar Raza and Isabella - 1
Sammay Haskel - 3

## 1. Merging

### 1.1. Reading the csvs

Read the `occupation.csv` and the `generation.csv` files into two dataframes. These files contain information about the `occupation` as well as the `generation` of several people.

In [2]:
df_occupation = pd.read_csv("C:\\Users\\samma\\Documents\\stat207\\saraza2\\lab_06\\occupation.csv")
df_occupation

first_name last_name            occupation
0       Tori     Kelly                singer
1       Tori   Ellison  statistics professor
2       Tori  Spelling               actress
3     Olivia   Rodrigo                singer
4     Olivia     Wilde               actress
5      Kelly   Findley  statistics professor

In [3]:
df_generation = pd.read_csv("C:\\Users\\samma\\Documents\\stat207\\saraza2\\lab_06\\generation.csv")
df_generation

first_name last_name  generation
0       Tori     Kelly  millennial
1       Tori   Ellison  millennial
2       Tori  Spelling       gen X
3     Olivia   Rodrigo       gen Z
4      Karle  Flanagan  millennial

### 1.2. Inner Join Merge

Suppose we clumsily decide to perform an **inner join merge** of the two dataframes on just the `first_name` column. Do this below and show the resulting dataframe.

In [7]:
df_primary = pd.merge(df_occupation, df_generation, how='inner', left_on=['first_name'], right_on=['first_name'])
df_primary

first_name last_name_x            occupation last_name_y  generation
0        Tori       Kelly                singer       Kelly  millennial
1        Tori       Kelly                singer     Ellison  millennial
2        Tori       Kelly                singer    Spelling       gen X
3        Tori     Ellison  statistics professor       Kelly  millennial
4        Tori     Ellison  statistics professor     Ellison  millennial
5        Tori     Ellison  statistics professor    Spelling       gen X
6        Tori    Spelling               actress       Kelly  millennial
7        Tori    Spelling               actress     Ellison  millennial
8        Tori    Spelling               actress    Spelling       gen X
9      Olivia     Rodrigo                singer     Rodrigo       gen Z
10     Olivia       Wilde               actress     Rodrigo       gen Z

### 1.3. Explaining the Result

How would you explain to a `STAT107 student` (who has never performed an **inner join merge** before) why the resulting merged dataframe in 1.2 resulted in 11 rows?

An inner join merge matches each primary key from the left dataframe with each primary key in the right dataframe. If the key matches then the merged row is included in the dataframe. If the key does not match, it is not included in the dataframe. As we only used first_name as the primary key, every first name from the left dataframe name matches with the same first name in the right dataframe, regardless of whether it is the correct person. This causes 11 rows.

### 1.4. Outer Join Merge

Suppose we also clumsily decide to perform an **outer join merge** of the two dataframes on just the `first_name` column. Do this below and show the resulting dataframe.

In [8]:
df_secondary = pd.merge(df_occupation, df_generation, how='outer', left_on=['first_name'], right_on=['first_name'])
df_secondary

first_name last_name_x            occupation last_name_y  generation
0        Tori       Kelly                singer       Kelly  millennial
1        Tori       Kelly                singer     Ellison  millennial
2        Tori       Kelly                singer    Spelling       gen X
3        Tori     Ellison  statistics professor       Kelly  millennial
4        Tori     Ellison  statistics professor     Ellison  millennial
5        Tori     Ellison  statistics professor    Spelling       gen X
6        Tori    Spelling               actress       Kelly  millennial
7        Tori    Spelling               actress     Ellison  millennial
8        Tori    Spelling               actress    Spelling       gen X
9      Olivia     Rodrigo                singer     Rodrigo       gen Z
10     Olivia       Wilde               actress     Rodrigo       gen Z
11      Kelly     Findley  statistics professor         NaN         NaN
12      Karle         NaN                   NaN    Flanagan  millennial

### 1.5. Explaining the Result

How would you explain to a `STAT107 student` (who has never performed an **outer join merge** before) why the resulting merged dataframe in 1.4 resulted in 13 rows?

An outer join merge matches each primary key from the left dataframe with each primary key in the right dataframe. If the key matches then the merged is included in the dataframe. If the key does not match, then both the unmatched rows are added into the final dataframe merged with NaN values. As we only used first_name as the primary key, every first name from the left dataframe name matches with the same first name in the right dataframe, regardless of whether it is the correct person. This causes 13 rows.

## 2. Linear Regression with Categorical Explanatory Variables

Suppose we like to build a linear regression model that predicts `number_of_Instagram_followers` given:
* `number_of_posts`
* `account_type`: personal, creator, business
* `have_profile_picture`: yes, no

Explain to a STAT107 student why we would seek to find the optimal intercept and slopes for THIS equation below...

\begin{align*}
\hat{\text{number\_of\_Instagram\_followers}} &= \hat{\beta}_0 \\
&+ \hat{\beta}_1 \times number\_of\_posts \\
&+ \hat{\beta}_2 \times account\_type_{personal} \\
&+ \hat{\beta}_3 \times account\_type_{creator} \\
&+ \hat{\beta}_4 \times have\_profile\_picture_{yes}
\end{align*}


... as opposed to THIS equation below...

\begin{align*}
\hat{number\_of\_Instagram\_followers} &= \hat{\beta}_0 \\
&+ \hat{\beta}_1 \times number\_of\_posts \\
&+ \hat{\beta}_2 \times account\_type_{personal} \\
&+ \hat{\beta}_3 \times account\_type_{creator} \\
&+ \hat{\beta}_4 \times account\_type_{business} \\
&+ \hat{\beta}_5 \times have\_profile\_picture_{yes} \\
&+ \hat{\beta}_6 \times have\_profile\_picture_{no}
\end{align*}


The first equation is better because it helps avoid multicollinearity and redundancy. By omitting one category for each categorical variable, we remove a category which is highly correlation with the rest of the categories helping us avoid multicollinearity. The model also becomes simpler to interpret as each coefficient represents the number of followers compared to the baseline category. The baseline category  is the category we dropped, in this case business and no.

## 3. Multicollinearity

The data in the `twitter_accounts.csv` file contains simulated `Twitter account` information for a given account including the:
* `number_of_followers` of the account
* `number_of_posts` the account has made
* `number_of_likes` the account has received on posts
* the number of `accounts_followed` by this particular account


### 3.1. Reading the csv
Read this csv file into a dataframe.

In [9]:
df_twitter = pd.read_csv("C:\\Users\\samma\\Documents\\stat207\\saraza2\\lab_06\\twitter_accounts.csv")
df_twitter

number_of_followers  number_of_posts  number_of_likes  accounts_followed
0                 2579.0            301.0            664.0              325.0
1                 2357.0            284.0            651.0              293.0
2                 2341.0            276.0            644.0              332.0
3                 2375.0            284.0            707.0              376.0
4                 2293.0            260.0            632.0              288.0
..                   ...              ...              ...                ...
995               2358.0            287.0            674.0              286.0
996               2554.0            308.0            674.0              390.0
997               2152.0            253.0            599.0              332.0
998               2053.0            259.0            650.0              271.0
999               2166.0            265.0            637.0              329.0

[1000 rows x 4 columns]

### 3.2. Train Test Split

Create a training and test dataset from this dataframe. Your training dataset should be comprised of a random sample of 80% of observations from your original dataset. Use a random state of `100`.

In [10]:
from sklearn.model_selection import train_test_split
df_train, df_test = train_test_split(df_twitter, test_size=0.2, random_state=100)
print(df_train.shape[0] / df_twitter.shape[0])
df_train.head()

0.8


number_of_followers  number_of_posts  number_of_likes  accounts_followed
675               1980.0            228.0            558.0              261.0
358               1969.0            229.0            567.0              316.0
159               2266.0            282.0            671.0              333.0
533               1928.0            234.0            543.0              239.0
678               2587.0            310.0            743.0              347.0

### 3.3. Multicollinearity

Suppose you'd like to build a linear regression model that predicts `number_of_followers` with:
* `number_of_posts`
* `number_of_likes`
* `accounts_followed`

#### 3.3.1 Full Model Multicollinearity

Explain to a STAT107 student why this model would suffer from multicollinearity.

In [12]:
df_train.corr()

number_of_followers  number_of_posts  number_of_likes  \
number_of_followers             1.000000         0.938795         0.883977   
number_of_posts                 0.938795         1.000000         0.940765   
number_of_likes                 0.883977         0.940765         1.000000   
accounts_followed               0.687315         0.647330         0.616689   

                     accounts_followed  
number_of_followers           0.687315  
number_of_posts               0.647330  
number_of_likes               0.616689  
accounts_followed             1.000000

In our linear regression model, number_of_flowers is the response variable where as the rest of the columns are the exlanatory variable. The reason this model will suffer from multicollinearity is because there is a high correlation (>0.7) between two explanatory variables. In this case, number_of_posts and number_of_likes have a high correlation (0.94).

#### 3.3.2. Reduced Model Lack of Multicollinearity

Explain to a STAT107 student why the model predicts `number_of_followers` with:
* `number_of_posts`
* `accounts_followed`


... does NOT have an issue with multicollinearity.

This model does not have an issue with multicollinearity because you have dropped the variable which is causing the high correlation between two explanatory variables. The remaining two variables do not have a high correlation with each other (<0.7) so they are a good fit for this model.

#### 3.3.3. Multicollinearity Downsides

Give two POTENTIAL downsides of using a linear regression model that has multicollinearity.

1. We will have misleading slope interpretations. The interpretation of some slopes becomes misleading and does not represent the actual relationship between the exlanatory and response variable.
2. Overfitting. By including both collinear explanatory pairs, it may increase the training dataset R^2 but decrease the test dataset R^2.